Frameworks/Foundation/_NSUndoManagerStack.mm

```diff
@@ -0,0 +1,47 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Starboard.h>
+#import "_NSUndoManagerInternal.h"
+#include <list>
+
+@implementation _NSUndoManagerStack {
+    std::list<StrongId<id<_NSUndoable>>> _stack;
+}
+
+- (void)pop {
+    if ([self count] > 0) {
+        _stack.pop_front();
+    }
+}
+
+- (id<_NSUndoable>)peek {
+    if ([self count] > 0) {
+        return _stack.front();
+    }
+    return nil;
+}
+
+- (void)push:(id<_NSUndoable>)undoObject {
+    _stack.push_front(undoObject);
+    [undoObject release];
```

Comment:
However, the same code is reused many many times. Removing this class would clutter the NSUndoManager. In addition, it makes sense to have the remove logic that will be added in this class as it deals strictly with the stack.

---
In reply to: [76338176](https://github.com/Microsoft/WinObjC/pull/900#discussion_r76338176) [](ancestors = 76338176,76294970,76160405)